### PR TITLE
What Matters In On-Policy Reinforcement Learning? A Large-Scale Empirical Study

### PR DESCRIPTION
**Pull request by AntoineTheb**

### CODE REVIEW

File: article/_posts/2020-07-02-What_Matters_in_RL.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+# Highlights
+
+- A large scale empirical study on design choices for Reinforcement Learning (RL) algorithms
+- Light recommendations are available from this study
+
+# Introduction
+
+> Deep reinforcement learning (RL) has seen increased interest in recent years due to its ability to have neural-network-based agents learn to act in environments through interactions. For continuous control tasks, on-policy algorithms such as REINFORCE, TRPO, A3C, PPO and off-policy algorithms such as DDPG and SAC have enabled successful applications such as quadrupedal locomotion, self-driving or dexterous in-hand manipulation.
+
+However, many of these algorithms employ high and low-level choices that may not be mentioned in their respective papers. Engstrom et al.[^1] showed that the original PPO paper failed to mention several key design choices that actually make up most of the improvements PPO had over prior methods. Code-level and algorithmic level hidden implementation details severly hinder reproducibility and slow down research in a field where reproducibility is already hard to achieve.[^2]
```
Comment (jhlegarreta) [inline article/_posts/2020-07-02-What_Matters_in_RL.md:23]: severly -> severely

File: article/_posts/2020-07-02-What_Matters_in_RL.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+However, many of these algorithms employ high and low-level choices that may not be mentioned in their respective papers. Engstrom et al.[^1] showed that the original PPO paper failed to mention several key design choices that actually make up most of the improvements PPO had over prior methods. Code-level and algorithmic level hidden implementation details severly hinder reproducibility and slow down research in a field where reproducibility is already hard to achieve.[^2] 
+
+# Methods
+
+The authors ran 250 000 experiments to investigate 68 implementation details separated in 8 different groups of choices: Policy Losses, Networks architecture, Normalization and clipping, Advantage Estimation, Training setup, Timesteps handling, Optimizers, and Regularization. The authors selected. For each group, the authors selected which choices might interact the most with each other, and left the remaining choices with default parameters that give competitive results.
```
Comment (jhlegarreta) [inline article/_posts/2020-07-02-What_Matters_in_RL.md:27]: Remove `The authors selected.` I assume it was finally included in the next sentence

File: article/_posts/2020-07-02-What_Matters_in_RL.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+# Remarks
+- A reader of this review wanting to follow these recommendations is STRONGLY encouraged to go read the actual paper before making decisions based on this review.
+- The paper is only 8 pages long but comes with 40 pages of appendix detailing the implementation details and experiment design
+- See also: [The 32 Implementation Details of Proximal Policy Optimization (PPO) Algorithm](https://costa.sh/blog-the-32-implementation-details-of-ppo.html)
+- This is only for on-policy algorithms, but off-policy algorithms also suffer from similar problems. Recurrent RL is also very much obscure to this day.
+
+# References
+
+[^1]: See [https://vitalab.github.io/article/2020/01/14/Implementation_Matters.html](https://vitalab.github.io/article/2020/01/14/Implementation_Matters.html) for more info
+[^2]: Joelle Pineau: Reproducibility, Reusability, and Robustness in Deep Reinforcement Learning ICLR 2018, https://www.youtube.com/watch?v=Vh4H0gOwdIg
```
Comment (jhlegarreta) [inline article/_posts/2020-07-02-What_Matters_in_RL.md:137]: Should be cross checked, but I think references need to have one blank line in between to be displayed on separate lines if that is the intention.

**Comment by AntoineTheb** [comment]
> Do they tell how they performed that analysis. It seems an interesting aspect.
Do you mean the choices themselves or how they were grouped ? The groups were chosen purely arbitrarily based on the suspicition of the authors so there's not much to expand upon. The choices themselves were selected by looking at popular RL codebases and prior work. Unfortunately there's not much more to it, but I'll clarify.

**Comment by jhlegarreta** [comment]
> Do you mean the choices themselves or how they were grouped ? The groups were chosen purely arbitrarily based on the suspicition of the authors so there's not much to expand upon. The choices themselves were selected by looking at popular RL codebases and prior work. Unfortunately there's not much more to it, but I'll clarify.

Yes, I was asking about how they chose or knew which parameters were interacting the most with each other. Your answer clarifies it. Thanks.

**Pull request merged**